sycl/doc/SYCLPluginInterface.md

```diff
@@ -0,0 +1,178 @@
+mentioned in this document. A part of PI API types and functions have exact
+matches in OpenCL. Whenever there is such a match, the semantics also fully
+matches unless the differences are explicitly specified in this document. While
+PI has roots in OpenCL, it does have many differences, and the gap is likely
+to grow, for example in the areas of memory model and management, program
+management.
+
+## Discovery and linkage of PI implementations
+
+![PI implementation discovery](images/SYCL_plugin_discovery.svg)
+
+Device discovery phase enumerates all available devices and their features by
+querying underlying plugins found in the system. This process is only performed
+once before any actual offload is attempted.
+
+### Plugin discovery
+
+Plugins are physically dynamic libraries stored somewhere in the system where
+the SYCL runtime runs. TBD - describe the process in details.
+
+#### Plugin binary interface
+TBD - list and describe all the symbols plugin must export in order to be picked
+up by the SYCL runtime for offload.
+
+#### OpenCL plugin
+
+OpenCL plugin is a usual plugin from SYCL runtime standpoint, but its loading
+and initialization involves a nested discovery process which finds out available
+OpenCL implementations. They can be installed either in the standard Khronos
+ICD-compatible way or not, and the OpenCL plugin can hook up with both.
+
+##### Linking to ICD-compatible OpenCL installations
+
+The plugin dynamically loads the OCL ICD loader - libOpenCL.so from system
+paths, and on success it asks the loader to provide all OpenCL implementations
+installed on the system. To do that it finds and calls the "clGetPlatformIDs"
+in the ICD loader and records the returned platform objects for later use. The
+plugin may choose to dynamically link with libOpenCL.so at build time.
+ 
+##### Linking to ICD-incompatible OpenCL installations
+
+For the OpenCL plugin to hook up such installation, it must be described in a
+configuration file similarly to plugin configuration file. *TBD - location and
+format of custom OpenCL installation config file*. The plugin will dynamically
+load the dynamic library pointed to in the configuration file and query it for
+OpenCL API function addresses similarly ICD loader. In more details, it will:
+- search for  the ```clGetExtensionFunctionAddress``` function in the 
+  library and call it with the string *“clIcdGetPlatformIDsKHR”* to find
+  the ```clIcdGetPlatformIDsKHR``` function pointer
+- then, using ```clIcdGetPlatformIDsKHR```, find the set of
+  ```cl_platform_id``` objects that the library provides, and store them
+  internally to enumerate available devices when required
+
+### Device enumeration by plugins
+
+TBD
+
+## PI API Specification
+
+PI allows for two levels of API implementation - basic and extended. The basic
+level comprises core APIs needed to perform SYCL offload. The extended level is
+needed only for OpenCL interoperability. 
```

Comment:
Just OpenCL or "device-specific runtime"?

Reply:
Let's follow the notation used in naming of PI interfaces (#222 and #259). There are core "pi" functionality, interoperability "picl" interfaces, and extended ones, those that are not in OpenCL but needed for SYCL.

Reply:
> Just OpenCL or "device-specific runtime"?

I think it can be later in the long term where we'd want better interop with other runtimes, but today it is OpenCL only. The other runtimes interop PI interfaces would probably be prefixed with "piext".